src/hupper/reloader.py

```diff
@@ -136,133 +130,178 @@ def run_once(self):
         )
-        return _run_worker(
-            worker,
-            self.monitor,
-            self.group,
-            self.logger,
-            self.reload_interval,
-            self.shutdown_interval,
-        )
+        return _run_worker(self, worker)
 
     def _wait_for_changes(self):
         worker = Worker(__name__ + '.wait_main')
-        return _run_worker(
-            worker,
-            self.monitor,
-            self.group,
-            SilentLogger(),
-            self.reload_interval,
-            self.shutdown_interval,
-        )
+        return _run_worker(self, worker, logger=SilentLogger())
+
+    @contextmanager
+    def _setup_runtime(self):
+        with self._start_control():
+            with self._start_monitor():
+                with self._capture_signals():
+                    try:
+                        yield
+                    except KeyboardInterrupt:
+                        pass
+
+    @contextmanager
+    def _start_control(self):
+        self.control_r, self.control_w = os.pipe()
+        try:
+            yield
+        finally:
+            os.close(self.control_r)
+            os.close(self.control_w)
+            self.control_r = self.control_w = None
 
+    def _control_proxy(self, signal):
+        return lambda *args: os.write(self.control_w, signal)
+
+    @contextmanager
     def _start_monitor(self):
-        proxy = FileMonitorProxy(self.logger, self.ignore_files)
+        proxy = FileMonitorProxy(
+            self._control_proxy(ControlSignal.FILE_CHANGED),
+            self.logger,
+            self.ignore_files,
+        )
         proxy.monitor = self.monitor_factory(
             proxy.file_changed,
             interval=self.reload_interval,
             logger=self.logger,
         )
         self.monitor = proxy
         self.monitor.start()
-
-    def _stop_monitor(self):
-        if self.monitor:
-            self.monitor.stop()
+        try:
+            yield
+        finally:
             self.monitor = None
+            proxy.stop()
 
+    _signals = {
+        'SIGHUP': ControlSignal.SIGHUP,
+        'SIGTERM': ControlSignal.SIGTERM,
+        'SIGCHLD': ControlSignal.SIGCHLD,
+    }
+
+    @contextmanager
     def _capture_signals(self):
-        # SIGHUP is not supported on windows
-        if hasattr(signal, 'SIGHUP'):
-            signal.signal(signal.SIGHUP, self._signal_sighup)
+        wrapped_signals = []
+        for signame, control in self._signals.items():
+            signum = getattr(signal, signame, None)
+            if signum is not None:
+                signal.signal(signum, self._control_proxy(control))
+                wrapped_signals.append(signum)
+        try:
+            yield
+        finally:
+            for signum in wrapped_signals:
+                signal.signal(signum, signal.SIG_DFL)
+
 
-    def _signal_sighup(self, signum, frame):
-        self.logger.info('Received SIGHUP, triggering a reload.')
-        self.monitor.set_changed()
```

Comment:
this old version was acquiring locks in the handler - now we just write to the control pipe